typing/ctype.ml

```diff
@@ -2127,7 +2488,11 @@ let get_gadt_equations_level () =
 (* a local constraint can be added only if the rhs
    of the constraint does not contain any Tvars.
    They need to be removed using this function *)
-let reify env t =
+(* [pre_id_pairs] describes the identifier substitutions that apply to the
+   type in the current environment, but for which the identifiers will not be
+   valid in the reified type.
+*)
+let reify env ?(pre_id_pairs = []) id_pairs t =
```

Comment:
There's a lot of new code in this function, and I'll confess to not having much looked yet at what it does. A priori I'm a bit surprised that we're complaining about scope escapes here; it might make sense, I don't know yet.

However, there seems to be nothing in the testsuite to exercise any part of this new code.
It would be nice if you could add some tests for it.

Reply:
I agree with @trefis that `reify` is not really the function to be checking for scope escapes.

I was all set to tell you that your example must already be handled by a call to `occur_unify` somewhere, since it can already happen with universal variables, and that call is the right place to emit the error.

However, it turns out that the equivalent case using universal variables is currently handled incorrectly: see #10189. Once that bug is fixed by adding a call to `occur_univar` somewhere, you should find your example fixed by adding the corresponding `occur_univar_or_unscoped` call.

Reply:
It isn't possible to use `occur_univar` for the example above, since it [introduces a genuine problem for univars](https://github.com/ocaml/ocaml/pull/10190#discussion_r571744889).

Instead, I've factored out the escape checking code to a dedicated `expanded_unscoped` function. Really, this function has different criteria than both `occur_univar` and `reify`, but in practise one or both of them are called in all of the cases where we need to check for these escapes. I've also added some comments, hopefully making it clearer that these checks only run where we provide the `pre_id_pairs` argument (and it `<> []`), and that we are only piggy-backing on their iteration over the type rather than fundamentally changing them.

I can break this out into a full dedicated iterator if it is still too unclear, and if it's acceptable to pay the cost of walking the type again.

Reply:
Your `expanded_unscoped` function seems to be trying to implement exactly the `unscoped` part of what `occur_univar_or_unscoped` should be doing. This should really just be part of `occur_univar_or_unscoped` as it has exactly the same purpose: something is about to be mutated, potentially allowing bound names to escape, ensure that the type which might escape has no bound names (or at least can be expanded such that it has no bound names).

Reply:
Also, unless I misunderstood Jacques reply to your comment, it doesn't seem to introduce a problem for univars.
(I'll let you or Leo confirm this)

Reply:
> Your `expanded_unscoped` function seems to be trying to implement exactly the `unscoped` part of what `occur_univar_or_unscoped` should be doing.

This is exactly as intended: it's also called by `occur_univar_or_unscoped`.

> Also, unless I misunderstood Jacques reply to your comment, it doesn't seem to introduce a problem for univars.

I think you understood perfectly, it was my sloppy writing above. His reply means that it's inappropriate to run the `occur_univar` part of `occur_univar_or_unscoped` [here](https://github.com/ocaml/ocaml/blob/3d083eedf9eb1de107b2790c794e883c9ac7554d/typing/ctype.ml#L2796). However, we need the `unscoped` part of it in order to catch the example I gave above.

Since `reify` walks over the type at that spot, I've hooked into it instead of walking over the type a second time.

Reply:
Could you merge the `expanded_unscoped` into `occur_univar_or_unscoped`. I think it will never make sense to call that function from anywhere else. For example, I don't think what you said above about the implications of Jacques reply is correct. Your patch should be calling `occur_univar_or_unscoped` exactly where Jacques' patch is calling `occur_univar` and you should not be calling `expanded_unscoped` anywhere else.

Reply:
Done.